sdk/go/pulumi/context.go

```diff
@@ -307,101 +299,89 @@ func (ctx *Context) RegisterResource(
-	outURN, resolveURN, rejectURN := NewOutput(nil)
-	urn := &resourceOutput{out: outURN, resolve: resolveURN, reject: rejectURN}
+func makeResourceState(custom bool, props map[string]interface{}) *ResourceState {
+	state := &ResourceState{}
+
+	state.urn = (*URNOutput)(newOutput(state))
 
-	var id *resourceOutput
 	if custom {
-		outID, resolveID, rejectID := NewOutput(nil)
-		id = &resourceOutput{out: outID, resolve: resolveID, reject: rejectID}
+		state.id = (*IDOutput)(newOutput(state))
 	}
 
-	state := make(map[string]*resourceOutput)
+	state.State = make(map[string]*Output)
 	for key := range props {
-		outState, resolveState, rejectState := NewOutput(nil)
-		state[key] = &resourceOutput{
-			out:     outState,
-			resolve: resolveState,
-			reject:  rejectState,
-		}
+		state.State[key] = newOutput(state)
 	}
 
-	return &resourceOutputs{
-		urn:   urn,
-		id:    id,
-		state: state,
-	}
+	return state
 }
 
 // resolve resolves the resource outputs using the given error and/or values.
-func (outputs *resourceOutputs) resolve(dryrun bool, err error, inputs map[string]interface{}, urn, id string,
+func (state *ResourceState) resolve(dryrun bool, err error, inputs map[string]interface{}, urn, id string,
 	result *structpb.Struct) {
-
 	var outprops map[string]interface{}
 	if err == nil {
 		outprops, err = unmarshalOutputs(result)
 	}
 	if err != nil {
 		// If there was an error, we must reject everything: URN, ID, and state properties.
-		outputs.urn.reject(err)
-		if outputs.id != nil {
-			outputs.id.reject(err)
-		}
-		for _, s := range outputs.state {
-			s.reject(err)
+		state.urn.s.reject(err)
+		if state.id != nil {
+			state.id.s.reject(err)
 		}
-	} else {
-		// Resolve the URN and ID.
-		outputs.urn.resolve(URN(urn), true)
-		if outputs.id != nil {
-			if id == "" && dryrun {
-				outputs.id.resolve("", false)
-			} else {
-				outputs.id.resolve(ID(id), true)
-			}
+		for _, o := range state.State {
+			o.s.reject(err)
 		}
+		return
+	}
 
-		// During previews, it's possible that nils will be returned due to unknown values.  This function
-		// determines the known-ed-ness of a given value below.
-		isKnown := func(v interface{}) bool {
-			return !dryrun || v != nil
-		}
+	// Resolve the URN and ID.
+	state.urn.s.resolve(URN(urn), true)
+	if state.id != nil {
+		known := id != "" || !dryrun
+		state.id.s.resolve(ID(id), known)
+	}
 
-		// Now resolve all output properties.
-		for k, s := range outputs.state {
-			v, has := outprops[k]
-			if !has && !dryrun {
-				// If we did not receive a value for a particular property, resolve it to the corresponding input
-				// if any exists.
-				v = inputs[k]
-			}
-			s.resolve(v, isKnown(v))
+	// During previews, it's possible that nils will be returned due to unknown values.  This function
+	// determines the known-ness of a given value below.
+	isKnown := func(v interface{}) bool {
+		return !dryrun || v != nil
```

Comment:
For my understanding, we assume that all values are known in non-dryrun execution? If a value is nil during a non-dryrun at this point, we assume that it will eventually resolve to something?

Reply:
A looks like I should've kept reading a few more lines. The next block seems to answer my question.